### ISSUE TITLE
statechage is triggered everytime pushState is called

### ISSUE THREAD

**Issue by matteomelani** [browserstate/history.js#96]
Probably I am just very confused....

so in Firefox 6.0.1 everytime I call pushState the statechange event get fired. Is this the expected behavior?

How do you go about restoring the state of your page?

For example let's say I have page users/1/post_board, the page presents to the user a list of posts. Now the user can simply filter the post by clicking a button. The button does an ajax call to /users/1/post_board?author=john. 
The app javascript invokes History.pushState() with a state object that contains all the info so that when the statechage event is triggered the page can be rebuilded. But since pushState trigger a statechage the page gets built twice. 

What am I missing here?
Do you have any example of how this should work? I am sorry to say but the demo is pretty useless since there is not state rebuilding in it.

Thanks!

**Comment by spyrbri** [comment]
+1

**Comment by R4wizard** [comment]
+1

**Comment by SuryaSankar** [comment]
+1
I am working around this. But this is definitely a huge bummer.

**Comment by errorx666** [comment]
:+1: Why is this behavior different from the spec?  Why do the `statechange` and `popstate` events both signal for the same condition (any change of state)?

**Comment by nigel32768** [comment]
+1
I had to use a global, I check it in the `statechange` event listener to see if it is being triggered a `pushstate` call or not.  

    gPushingHistoryState = true;
    History.pushState(x, y, z);
    gPushingHistoryState = false;

Is there any way to tell from the `statechange` event whether it was initiated by a back button or a `pushstate` call?

**Comment by danger89** [comment]
Yep, there is an issue here. I like to get it solved in the master. Eg. using this:
https://github.com/browserstate/history.js/pull/401/files

It's still not fixed.

**Comment by webarthur** [comment]
The solution:

```
(function(history){
    var pushState = history.pushState;
    history.pushState = function(state) {
        if (typeof history.onpushstate == "function") {
            history.onpushstate({state: state});
        }
        // ... whatever else you want to do
        // maybe call onhashchange e.handler
        return pushState.apply(history, arguments);
    }
})(window.history);
```

from: http://felix-kling.de/blog/2011/01/06/how-to-detect-history-pushstate/
